find feature function

ahmedtarek26 opened the pull request:
Instead of using loops, this function is to search for a specific feature in the sbol3 file.
I added the function in the `helper_function.py`, I think this is a good place to put this function, also we can add the function in the community media repository.

ahmedtarek26 committed:
find_feature function #185

This function is to search for a feature in the sbol3 file

jakebeal commented on sbol_utilities/helper_functions.py:
```diff
@@ -357,3 +357,9 @@ def collector(obj: sbol3.Identified):
         doc.traverse(collector)
     return outgoing
+
+def find_feature(sbol3: sbol3_file,f: feature,seq: sequence) -> sbol3.Feature:
```
Can you please add a docstring that documents the behavior of this function, following the same pattern as the other functions in the file?

It also appears that your type hints are incorrect; does this run correctly when you test it?

jakebeal commented on sbol_utilities/helper_functions.py:
```diff
@@ -357,3 +357,9 @@ def collector(obj: sbol3.Identified):
+def find_feature(sbol3: sbol3_file,f: feature,seq: sequence) -> sbol3.Feature:
+    for i in sbol3.features:
+        if i.instance_of == f.identity:
+            i.locations.append(Range(seq, 746, 875))
```
Why are you appending the Range here?

jakebeal commented:
This function needs to have a test added that validates its behavior before it can be merged.

ahmedtarek26 committed:
implement find_ feature and test_find_feature

I implemented `find _feature` and `test_find_feature` functions.

I removed the range and the function now could check if the input feature exist on the input component or not

jakebeal commented on sbol_utilities/helper_functions.py:
```diff
@@ -357,3 +357,12 @@ def collector(obj: sbol3.Identified):
+def find_feature(doc: sbol3.Component,f:sbol3.Feature) -> sbol3.Feature:
+    """
+    find_feature function used for  finding a specific feature you specify and check
+    if this feature one of your sbol3 document features or not.
```
It is not a document, but a component. You should change the variable name as well.

jakebeal commented on test/test_helpers.py:
```diff
@@ -136,6 +136,29 @@ def test_outgoing(self):
+        i13504.features.append(sbol3.SubComponent(b0034))
+
+        result=find_feature(i13504,b0034)
+        self.assertEqual(find_feature(i13504,b0034),result)
```
This test does not actually check if the function is operating correctly. How would you know if result is actually the right result?

jakebeal commented on sbol_utilities/helper_functions.py:
```diff
@@ -357,3 +357,12 @@ def collector(obj: sbol3.Identified):
         doc.traverse(collector)
     return outgoing
+
+def find_feature(doc: sbol3.Component,f:sbol3.Feature) -> sbol3.Feature:
```
This isn't actually addressing #185, as one has to already have the feature in order to find it.
In the issue, the example is looking for a feature whose identity is known, but not that actual object.

ahmedtarek26 replied:
yeah, got it

what if I used this line of code? 
` if i.instance_of.split('/')[-1] == f.identity.split('/')[-1]:`
using `.identity` with the feature we input and find the feature name then comparing it with the same feature if it exist using `.instance_of`

jakebeal replied:
That doesn't address the question of the function's interface.

jakebeal commented on sbol_utilities/helper_functions.py:
```diff
@@ -357,3 +357,12 @@ def collector(obj: sbol3.Identified):
+    if this feature one of your sbol3 document features or not.
+    """
+    for i in doc.features:
+        if i.instance_of == f.identity:
```
This will never return true.

jakebeal commented on sbol_utilities/helper_functions.py:
```diff
@@ -357,3 +357,12 @@ def collector(obj: sbol3.Identified):
+    """
+    for i in doc.features:
+        if i.instance_of == f.identity:
+            return i
```
What should happen if you don't find the feature?

ahmedtarek26 committed:
Update find_feature and test_find_feature

Adding expected values in test_find_feature function and make some changes in find_feature function

ahmedtarek26 committed:
Merge branch 'find_feature' of https://github.com/ahmedtarek26/SBOL-utilities into find_feature

ahmedtarek26 committed:
Update find_feature function

tcmitchell commented:
@jakebeal Please note that this PR is targeted to `main` instead of `develop`. That might not be what you want.

jakebeal commented:
As no updates have been made within the last several months to address the requested changes, I am going to close this PR.
